File: surf-chat/src/main/java/dev/slne/surf/social/chat/command/argument/ChannelArgument.java
package dev.slne.surf.social.chat.command.argument;

import dev.jorel.commandapi.arguments.ArgumentSuggestions;
import dev.jorel.commandapi.arguments.CustomArgument;
import dev.jorel.commandapi.arguments.StringArgument;
import dev.slne.surf.social.chat.object.Channel;
import dev.slne.surf.social.chat.provider.ChannelProvider;
import java.util.stream.Collectors;

public class ChannelArgument extends CustomArgument<Channel, String> {
  public ChannelArgument(String nodeName) {
    super(new StringArgument(nodeName), info -> {
      Channel channel = Channel.getChannel(info.input());

      if (channel == null) {
        throw CustomArgumentException.fromMessageBuilder(new MessageBuilder("Unknown channel: ").appendArgInput());
      } else {
        return channel;
      }
    });

    this.replaceSuggestions(ArgumentSuggestions.strings(info -> ChannelProvider.getInstance().getChannels().values().stream().map(Channel::getName).toArray(String[]::new)));
  }
}

File: surf-chat/src/main/java/dev/slne/surf/social/chat/command/argument/ChannelArgument.kt
package dev.slne.surf.social.chat.command.argument

import dev.jorel.commandapi.arguments.ArgumentSuggestions
import dev.jorel.commandapi.arguments.CustomArgument
import dev.jorel.commandapi.arguments.StringArgument
import dev.slne.surf.social.chat.`object`.Channel
import dev.slne.surf.social.chat.provider.ChannelProvider

class ChannelArgument(nodeName: String) : CustomArgument<Channel, String>(
    StringArgument(nodeName),
    { info ->
        Channel.getChannel(info.input()) ?: throw CustomArgumentException.fromMessageBuilder(
            MessageBuilder("Unknown channel: ").appendArgInput()
        )
    }) {
    init {
        replaceSuggestions(ArgumentSuggestions.stringCollection { ChannelProvider.channels.values.map { it.name } })
    }
}